task/src/main/java/com/flipkart/flux/impl/task/AkkaTask.java

```diff
@@ -77,121 +76,127 @@
+
+                    } catch (HystrixRuntimeException hre) {
+                        FailureType ft = hre.getFailureType();
+                        // we signal a timeout for any of Timeout, ThreadPool Rejection or Short-Circuit - all of these may go through with time and retry
+                        if (ft.equals(FailureType.REJECTED_THREAD_EXECUTION) || ft.equals(FailureType.SHORTCIRCUIT) || ft.equals(FailureType.TIMEOUT)) {
+                            // update flux runtime with task outcome as timeout
                             fluxRuntimeConnector.updateExecutionStatus(new ExecutionUpdateData(
                                     taskAndEvent.getStateMachineId(), taskAndEvent.getTaskId(), Status.errored, taskAndEvent.getRetryCount(),
-                                    taskAndEvent.getCurrentRetryCount(), cause.getMessage()));
-                            throw new FluxError(FluxError.ErrorType.retriable, cause.getMessage(), null, false,
+                                    taskAndEvent.getCurrentRetryCount(), "Execution timeout for : " + task.getName()));
+                            throw new FluxError(FluxError.ErrorType.timeout, "Execution timeout for : " + task.getName(),
+                                    null, false,
                                     new FluxError.ExecutionContextMeta(taskAndEvent.getStateMachineId(), taskAndEvent.getTaskId(),
                                             taskAndEvent.getRetryCount(), taskAndEvent.getCurrentRetryCount()));
                         } else {
-                            // mark the task outcome as execution failure
-                            fluxRuntimeConnector.updateExecutionStatus(
-                                    new ExecutionUpdateData(taskAndEvent.getStateMachineId(), taskAndEvent.getTaskId(), Status.errored,
-                                            taskAndEvent.getRetryCount(), taskAndEvent.getCurrentRetryCount(), hre.getMessage()));
+                            //check if the exception hierarchy has FluxRetriableException, if yes trigger retry
+                            boolean isFluxRetriableException = false;
+                            Throwable cause = hre.getCause();
+                            Throwable rootCause = cause;
+                            while (cause != null) {
+                                if (cause.getClass().getName().equals(FluxRetriableException.class.getName())) {
+                                    isFluxRetriableException = true;
+                                    break;
+                                }
+                                rootCause = cause;
+                                cause = cause.getCause();
+                            }
+                            if (isFluxRetriableException) {
+                                // mark the task outcome as execution failure
+                                fluxRuntimeConnector.updateExecutionStatus(new ExecutionUpdateData(
```

Comment:
once we have 1 variable to track cause, updateExecState on line 141, 149 can be moved out of if-else block.